apicast/src/policy/find_service.lua

```diff
@@ -0,0 +1,73 @@
+local next = next
+
+local policy = require('policy')
+local _M = policy.new('Find Service Policy')
+local configuration_store = require 'configuration_store'
+local new = _M.new
+
+local function find_service_strict(configuration, host)
+  local found
+  local services = configuration:find_by_host(host)
+
+  for s=1, #services do
+    local service = services[s]
+    local hosts = service.hosts or {}
+
+    for h=1, #hosts do
+      if hosts[h] == host and service == configuration:find_by_id(service.id) then
+        found = service
+        break
+      end
+    end
+    if found then break end
+  end
+
+  return found or ngx.log(ngx.WARN, 'service not found for host ', host)
+end
+
+local function find_service_cascade(configuration, host)
+  local found
+  local request = ngx.var.request
```

Comment:
Should this module receive the request via `context` ?

Reply:
Interesting idea. Guess that depends if changes like `req.set_uri` or `set_method` are propagated and are changing this variable. If yes, then policies can change this variable indirectly and it is all fine. But if this is immutable then it should be in the context, so other policies can override it.
And maybe it is just better to get rid of this variable and use more granular nginx apis. Because it just splits the string into method, path, etc. anyway https://github.com/3scale/apicast/blob/master/apicast/src/configuration.lua#L201-L204 so it can use req.get_method etc.